### ISSUE TITLE
yarn_install task is not installing devDependencies

### ISSUE THREAD

**Issue by nikolalsvk** [rails/webpacker#1212]
Due to the recent changes that have been made, `webpacker:yarn_install` task is not installing `devDependencies`. This is breaking my CI builds, since I relied on webpacker to do the install of packages for my test environment on the CI server.

Change that brought this in the webpacker gem https://github.com/rails/webpacker/commit/d5fe0fe8e492d1511c6f31648f8fe861504c111e#diff-5a0f30530a2f1282593526bb85d65029
Change that brought this in the Rails gem
https://github.com/rails/rails/commit/2118c98b98c4984069e070aa0d5c398de618d02a#diff-dcd44149ba1cd25c12956a3ebecc32a3

I'm just wondering if we could introduce an optional `--production=true|false` flag in the rake task that does `yarn install`?

Or create a `NODE_ENV` dependant flag for the rake task?

**Comment by gauravtiwari** [comment]
@nikolalsvk Perhaps you can use yarn directly instead of relying on rake task?

**Comment by nikolalsvk** [comment]
@gauravtiwari yes, I'm doing `yarn install` after assets precompilation, but that is just one extra `yarn install` to do.

I'm just wondering could there be a different yarn_install task for testing environment for example?

**Comment by gauravtiwari** [comment]
@nikolalsvk I guess what you could do is disable webpacker rake task and create a new one as before: 

```rb
# lib/tasks/yarn_install.rake
Rake::Task['yarn_install'].clear

task 'yarn_install' do
    # old task here
end
```

Would this work?

**Comment by nikolalsvk** [comment]
@gauravtiwari guess it would work, thanks.

I'm just curious why not leave the choice whether to install or not to install devDependencies?

**Comment by cipater** [comment]
I'm having issues with this change as well. In particular, `yarn install --production` seems to be removing the babel-plugin-relay, but it's required as part of the asset precompilation... which is now failing due to being unable to find the plugin.

**Comment by arthow4n** [comment]
IMHO lots of people install `webpack-*` and `babel-*` in devDependencies since they are for building instead of serving things. `yarn install --production` can cause devDependencies being used to build removed. I always perform `yarn` before doing `assets:precompile` but now I get my `webpack-chunk-hash` in devDependencies missing after `webpacker:yarn_install` after upgrading webpacker. I don't think it is a good idea forcing `yarn install --production` in webpacker rake task.

**Comment by catskull** [comment]
Maybe I'm doing something wrong, but this change broke my production deploys. I too had all my compilation related packages in devDependencies, so the production server couldn't build the assets correctly.

**Comment by odlp** [comment]
Cross-linking to #1330, where the task ends up removing dev dependencies which have already been installed.

**Comment by luccasmaso** [comment]
Any solutions?

**Comment by groksrc** [comment]
I am also having issues with `--production` being hardcoded. I have a lot of things in `devDependencies` that don't belong in production and I am more likely to remove webpacker than I am to change this. I also have more than just production and development environments and would like more flexibility in passing options to NODE_ENV. It's fine if webpacker makes certain assumptions for me but it shouldn't force me to do things a certain way.

**Comment by leikind** [comment]
I have a Rails 5.1.6 app with webpacker successfully deployed on Heroku app with 2 buildpacks: 
(1)`nodejs` and (2) `ruby`.

Upgrading to 5.2.0 breaks the Heroku build for the reasons @nikolalsvk listed above. The build fails because there is a number of `devDependencies` not installed after `rake yarn:install`.

Moving `devDependencies` to `dependencies` feels very very wrong.

What is interesting is Heroku's approach to handling `devDependencies` and `dependencies` in their `nodejs` buildpack:  

> By default, Heroku will install all dependencies listed in package.json under dependencies and devDependencies.

> After running the installation and build steps Heroku will strip out the packages declared under devDependencies before deploying the application.

( https://devcenter.heroku.com/articles/nodejs-support#package-installation )

It seems like a more reasonable approach than just excluding all `devDependencies`.

Unfortunately, the rails buildpack is executed _after_ the nodejs buildpack, and `webpacker:yarn_install` makes sure none of my `devDependencies` are present.

I am wondering what the correct solution should be.

**Comment by gauravtiwari** [comment]
@leikind They are not `devDependencies` really because you are using them on Heroku as well, which is a production environment. If however, you compile your assets locally and then push them to Heroku then they can safely live under dev dependencies, considering `assets:precompile` is never run on Heroku. 

With webpacker, we recommend putting all build and app related dependencies under dependencies section in package.json (except test deps). I understand the confusion but you see they are not app dependencies i.e. it won't be included in your pack but they are needed to produce a pack or build - either on Heroku or on your local machine. Does this make sense?

**Comment by leikind** [comment]
The app is indeed built on Heroku, `assets:precompile` is executed. 

You see, if my app were a Node.js app using only the node buildpack, I'd have no issues building my app on Heroku, as the node buildpack would install all packages first, run build scripts (webpack for instance), and then strip devDependencies. So, the node team in Heroku does invite its customers to build their node apps on Heroku.

I see no argument against building my app on Heroku except for the webpacker team making it harder.

Here's my current hack to solve my problem:

```ruby
namespace :yarn do
  desc 'Install all JavaScript dependencies as specified via Yarn'
  task :install do
    system('./bin/yarn install --no-progress')
    Rake::Task['yarn:install'].clear
  end
end
```

**Comment by mhfen** [comment]
> @leikind They are not `devDependencies` really because you are using them on Heroku as well, which is a production environment. If however, you compile your assets locally and then push them to Heroku then they can safely live under dev dependencies, considering `assets:precompile` is never run on Heroku.

Many teams run assets:precompile on heroku. In fact, it is a standard part of the rails build pack...

> With webpacker, we recommend putting all build and app related dependencies under dependencies section in package.json (except test deps). I understand the confusion but you see they are not app dependencies i.e. it won't be included in your pack but they are needed to produce a pack or build - either on Heroku or on your local machine. Does this make sense?

It does not. Webpack, babel, etc are dev ops tools. They are not application code. Heroku installs devDeps, so why force everyone to make a bad organization decision and move those to deps?

This should be addressed by the Webpacker team...

**Comment by richseviora** [comment]
> With webpacker, we recommend putting all build and app related dependencies under dependencies section in package.json (except test deps). I understand the confusion but you see they are not app dependencies i.e. it won't be included in your pack but they are needed to produce a pack or build - either on Heroku or on your local machine. Does this make sense?

@gauravtiwari I agree this makes sense if you look at the `devDependencies` / `dependencies` distinction as code that's used only locally during the dev workflow (like Prettier, ESLint, etc) vs code that's executed on production. I think what you're hearing from other people (and please count myself in that list) is that they all have different understandings of the term and some of those understandings have been implemented into widely used tools (like Heroku).

I ran into the same problem myself when deploying this via the `opsworks_ruby` cookbook to AWS. My solution was to move the `devDependencies` into `dependencies`, but I'd like to see it install dependencies preferably by default, or as an option.

**Comment by Zyphrax** [comment]
This is really confusing. I have the following devDependencies:
```json
"devDependencies": {
  "@types/bootstrap": "<4",
  "@types/braintree-web": "^3.6.3",
  "@types/chartist": "^0.9.43",
  "@types/daterangepicker": "<3",
  "@types/dropzone": "^5.0.4",
  "@types/jquery": "<3",
  "@types/jqueryui": "^1.12.6",
  "@types/moment-timezone": "^0.5.9",
  "caniuse-lite": "^1.0.30000910",
  "node-sass": "^4.10.0",
  "resolve-url-loader": "^3.0.0",
  "sass-loader": "^7.1.0",
  "ts-loader": "^5.3.0",
  "tslint": "^5.11.0",
  "typescript": "^3.1.6",
  "webpack": "^4.26.1",
  "webpack-bundle-analyzer": "^3.0.3",
  "webpack-dev-server": "^3.1.10"
}
```

It doesn't make sense to me to move all the TypeScript typings and Webpack loaders in dependencies just to make compilation work. 

Would it be possible to remove the Yarn install step?

**Comment by jakeNiemiec** [comment]
>@leikind @Zyphrax It doesn't make sense to me to move all the TypeScript typings and Webpack loaders in dependencies just to make compilation work.

`devDependencies` has more usefulness in the context of a JS library.

Let's take [`webpack`](https://github.com/webpack/webpack/blob/master/package.json#L70) as an example. Notice that `react` is installed under `devDependencies`. 

- If you developing for the [`webpack`](https://github.com/webpack/webpack/blob/master/package.json#L70) library locally, it would install `react`.
- If you developing for a rails app that lists [`webpack`](https://github.com/webpack/webpack/blob/master/package.json#L70) as a dependancy, it would **NOT**  install `react`. This is because webpack does not need `react` to function, therefore it does not need to pass the lib _downstream_. 

This is what @gauravtiwari was getting at. Your rails app is "the end of the JS line", you don't need to worry about downstream users. I am aware that many READMEs will tell you to use `--save-dev`, but `devDependencies` offers no practical use. (esp. in the context of webpacker)

**TL;DR:** If you need something to build you production bundle, it cannot be placed in `devDependencies`. (including: webpack, typescript, node-sass, sass-loader)

**Comment by chevinbrown** [comment]
@jakeNiemiec is there some confusion here about what `devDependencies` vs production dependencies means? I feel like many believe that the `dependencies` include something in a production bundle and therefore increase the package size? (if this question/comment is just chatter, thumbs down and I'll delete)

**Comment by odlp** [comment]
IMO there are legitimate use cases for `devDependencies` in a Rails app - say you want to have JS testing libraries like Karma, Jasmine etc installed via `yarn` on CI but don't want these when in the production environment (see https://github.com/rails/webpacker/issues/1330 for further context).

Webpacker hardcoding `--production` into a Rake task is disruptive of other `yarn` uses which don't relate to asset compilation.

**Comment by jakeNiemiec** [comment]
> @odlp IMO there are legitimate use cases for devDependencies in a Rails app

You _never_ need to worry about downstream `dependencies` in a Rails app, if you come across a `--save-dev` in a README, that was not put there for you (the same problem exists in Node.js-land). @nikolalsvk, What is the purpose for a CI test? To simulate a production environment and programmatically test a bunch of edge cases. Wouldn't you want to test against minified/transpiled production bundles?

> say you want to have JS testing libraries like Karma, Jasmine etc installed via yarn on CI but don't want these when in the production environment (see #1330 for further context).

Are you using webpack to bundle Karma or Jasmine? I don't understand why you would do this. 

> @chevinbrown Is there some confusion here about what devDependencies vs production dependencies means?  I feel like many believe that the dependencies include something in a production bundle and therefore increase the package size? 

Less "confusion" and more like ritual inclusion of programing patterns that subvert their original purpose (e.g.: running CI tests in development mode). All it does is tell yarn/npm what libraries to put in `node_modules`. If you include Karma in `dependencies`, you will just additionally have Karma and it's `dependencies` in your `node_modules`. [At worse you have an extra 8 seconds to wait when you install your `dependencies`.](https://bundlephobia.com/result?p=karma@3.1.1)

TL;DR: You should follow what @gauravtiwari said, **there are [mechanical differences between the `production` and `development`](https://webpack.js.org/concepts/mode/) flags**, they are more than just labels:
![image](https://user-images.githubusercontent.com/16782063/49112639-e9e5de80-f258-11e8-98fc-7b29d01ec219.png)

**Comment by odlp** [comment]
@jakeNiemiec sorry if I was unclear, but to clarify:

> Are you using webpack to bundle Karma or Jasmine? I don't understand why you would do this.

I'm using `yarn` to install Karma and Jasmine. `webpacker` also happens to use `yarn` to install dependencies, but by hardcodes `--production` in it's rake tasks which interferes with non-weback related tasks.

My suggestion is the webpacker commands should be environment aware.

**Comment by jakeNiemiec** [comment]
> @odlp My suggestion is the webpacker commands should be environment aware.

![image](https://user-images.githubusercontent.com/16782063/49165040-f2402700-f2f5-11e8-9340-e6b7ac9e0f02.png)

- Think of it like this, if you really wanted to, you could hit a nail with the non-flat side of a hammer...it would work, but you would be more prone to problems.

- In the same way setting `NODE_ENV` **comes with side-effects specific to your `dependencies`** and the `dependencies` of those `dependencies`.

If you pass production bundles to your tests, you will get **less surprises**, **faster testing times**, and **results that are truer to your actual production env**.

Real-life example: [The Font Awesome package tends to break when using webpack 4, **but only if you are in `production`** mode.](https://fontawesome.com/how-to-use/with-the-api/other/tree-shaking#issues)

_However_, if you really want to have the autonomy to do this, you can use this trick to force a different environment: `NODE_ENV=development karma start ./config/karma.conf.js` or however you are running your tests, just set or unset NODE_ENV before the relevant command. If you are still missing karma from your CI server, [you can globally add it with: `yarn global add karma`](https://yarnpkg.com/lang/en/docs/cli/global/). If you do use this, please note that this has its own perils and required reading. Best of luck!

**Comment by trent-boyd** [comment]
@gauravtiwari

> With webpacker, we recommend putting all build and app related dependencies under dependencies section in package.json (except test deps).

That's an example of bikeshedding if I ever saw one. Further, this is listed _nowhere_ in the documentation. I should not have to crawl the Webpacker source or track down this exact issue just to find out that your team is imposing their own view on how `devDependencies` should work.

Hundreds of other JS-based open source projects intentionally instruct you to use the `--dev` flag during installation. Webpacker is a _development tool_ for bundling up code intended to be run _in a browser_. It would make sense to me that other development tools (like Babel, its plugins, and other Webpack loaders) would also be installed as _development dependencies_.

And besides that, what is the `--production` flag really helping anyway? If I want to really streamline my NPM install times or cut down on disk usage, that should be something I opt-in to.

**Comment by ytbryan** [comment]
> I **should** not have to crawl the Webpacker source or track down this exact issue just to find out that your team is imposing their own view on how devDependencies should work.

@trent-boyd Mr Boyd, please feel free to send a PR 

Thank you @gauravtiwari for your time contributing to the rails/webpacker project. Webpacker has been a delight and it saves me so much time integrating Vue into Rails <3

**Comment by trent-boyd** [comment]
I don't think you'd like the the voice of my documentation on this one. 😉

It's just frustrating to see a lack of attention to his, especially when dev team has [come up against it themselves](https://github.com/rails/webpacker/issues/117) and there's been an [unaddressed PR](https://github.com/rails/webpacker/pull/1331) about it for nine months. Even closing something as a #wontfix is fine if that's what the maintainers decide.

**Comment by jakeNiemiec** [comment]
> @trent-boyd It's just frustrating to see a lack of attention to his, especially when dev team has come up against it themselves 

The dev team has come up against it? Pretty sure the resolution was the opposite of what you portray: 
> [Yes let's fix the root issue and **move away from errant use of devDependencies**](https://github.com/rails/webpacker/issues/117#issuecomment-282798638).

> That's an example of bikeshedding if I ever saw one. I should not have to [...] find that your team is imposing their own view on how devDependencies should work.

👆That's bikeshedding if _I_ ever saw it. The WebpackER team is aligning themselves with how Webpack is used by the Node.js community. **This is smart**, it will do a lot to reduce friction between Rails & JS pipelines. 

Remember that `package.json` & `devDependencies` are all constructs from the Node.js world. Nothing is being imposed, it's just being aligned with [existing usage/documentation](https://yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-production-true-false).

> Hundreds of other JS-based open source projects intentionally instruct you to use the --dev flag during installation. 

Like I said [here](https://github.com/rails/webpacker/issues/1212#issuecomment-442228316), those instructions aren't always meant for end users that are not creating a package themselves (like any Webpacker app). The 2 links below are useful for explaining why this is:
https://yarnpkg.com/blog/2018/04/18/dependencies-done-right/
https://stackoverflow.com/a/22004559/6091685

> Webpacker is a development tool for bundling up code intended to be run in a browser. It would make sense to me that other development tools (like Babel, its plugins, and other Webpack loaders) would also be installed as development dependencies.

Rails is a development tool too 🙃. Did you know that both [Webpack](https://webpack.js.org/concepts/mode/) and [Babel are set up to behave differently depending on which flag is set? (NODE_ENV)](https://babeljs.io/docs/en/config-files#apicache) I personally would want to run tests against production bundles.

**The webpacker team is just trying to keep you on the most pain-free path**. Just chuck your `devDependencies` in `dependencies` and be done with it...[none of this will matter soon anyway because the rules are changing ¯\\_(ツ)_/¯ ](https://github.com/yarnpkg/rfcs/files/2378943/Plugnplay.pdf).

**Comment by trent-boyd** [comment]
> Yes let's fix the root issue and move away from errant use of devDependencies.

Neat. Awesome. I'm on board as long as it's in the docs.

> Did you know that both Webpack and Babel are set up to behave differently depending on which flag is set? (NODE_ENV). I personally would want to run tests against production bundles.

Yes, and that's how I'd prefer it to stay. [Using](https://github.com/rails/webpacker/blob/master/lib/tasks/webpacker/yarn_install.rake#L4) the `--production` flag bypasses that. And the test guy is someone else, but I do agree with you on that.

**Comment by nikolalsvk** [comment]
We're still having production flag hardcoded y'all and I still don't understand why :smile:

**Comment by csalvato** [comment]
Reading this thread, I'm finding myself super confused.  Are we supposed to not use devDependencies at all when using Webpacker? Is that the intended use? 😕 

We don't use this rake task in our project (we just run `yarn install` directly), but found that [using devDependencies were causing us problems](https://github.com/rails/webpacker/issues/1395#issuecomment-452856427) when debugging a discrepancy between dev and prod environments

**Comment by catskull** [comment]
We just don't use devDependencies. Yeah extra packages might get downloaded in production but it shouldn't be that big of a deal. To me it's just a paper cut, but the few hours I spent debugging I'll never get back.

**Comment by jakeNiemiec** [comment]
@csalvato Don't feel bad, people have been confusing this for over a year:

![image](https://user-images.githubusercontent.com/16782063/50930474-b8e6f500-1425-11e9-9118-555ca68c4fec.png)
